### DISCUSSION TITLE
Multi-turn / session experiments in datasets

### DISCUSSION THREAD

**Discussion by bartsimo** [langfuse/langfuse#4208] [Ideas]
Hi,

with the Datasets feature, is the creation of multi-turn items supported, e.g. in a RAG case:

Turn 1:
Human: Hi, can you please give me data on X?
AI: < Gives data about X >

Turn 2:
Human: Please focus on the years 2020-2021
AI: <Gives filtered data from Turn 1>

For up to n turns.

From what I can see in the documentation and the cookbook, only single-turn is shown.
https://langfuse.com/docs/datasets/python-cookbook

https://langfuse.com/docs/datasets/overview

Can experiments that contain multi-turn interactions be automated with Datasets?

**Comment by jannikmaierhoefer** [comment]
Hi @bartsimo, this is not yet supported. I, therefore, converted this post into an "idea" post.

**Reply by marcklingen** [reply]
Quick addition:

As langfuse dataset runs can be linked to traces, you can run an **integration test** against a multi-step RAG application or agent. Each dataset item will then be linked to a single execution trace of the whole system. You can add evaluators for every step of your application and add them as `scores` to langfuse.

Alternatively, we often see teams splitting these chains into **unit tests** which they test separately with different datasets. Thereby they keep all other variables constant and can optimize the different steps independently. This also applies to testing single turns within a conversation separately to make the tests more robust.

Sessions (multiple-invocations) for each dataset item are currently not supported natively. You can however do them all within one trace if you want. Linking sessions to datasets would be interesting, thank you for raising this.

**Reply by bartsimo** [reply]
Thank you both very much for your quick replies!

Our main focus is on test automation. In the tests, we are not so much interested in asserting expected output nor scores. We are interested in the interactive part of the RAG chatbot, which is currently assessed qualitatively.
90% of what we are interested in are multi-turn interactions:
- How well is context remembered
- How fluent is the conversation
- Are guardrails honored mid-turn 
- ...

In our case, my example
Turn 1:
Human: Hi, can you please give me data on X?
AI: < Gives data about X >

Turn 2:
Human: Please focus on the years 2020-2021
AI: <Gives filtered data from Turn 1>

Would be ONE test case in which we do not want to split up the turns in any way since we are interested in the joined qualitative performance.

@marcklingen 
If I cannot create multi-turn items, I do not know how I should automate such testing in Langfuse?
Do your additions apply, if this is our testing objective?
I could not quite map your addition to this.

Thank you guys so much!
The traces are awesome!

**Reply by marcklingen** [reply]
Have you tried running all turns within a single trace when in development? Thereby you can map one multi-turn interaction to a single dataset item and then evaluate the interaction

**Comment by feynnon** [comment]
I was looking for the same thing in the docs - may I suggest to specifically point this out in your docs? Maybe one page on the topic of multi-turn conversations, and that the UI does not natively support it? Would have saved me a lot of time, until I finally found this. I'm likely not alone.
Thanks for langfuse btw, it's pretty nice!

**Reply by wbreilly** [reply]
Seconding this. Our product is all about whole conversations, which are logged as sequences of traces. Our engineers need this setup, but our domain experts need to be able to annotate sessions. 

Please build support for session annotation, session evals with LLM as judge, and session annotation queues. Thank you.

**Comment by judsoniv** [comment]
+1 for the ability to run LLM-as-judge evals on entire conversations. Like others here, we are interested in the overall performance profile of an extended conversation, and we have metrics we would like to evaluate against an entire conversation. We care less about the perfect response to a particular question and instead care about the tradeoffs between different aspects of the conversation that we track. Essentially, the sum of evals of individual traces do not necessarily equal how we would evaluate the conversation as a whole.

It would be good to know if out-of-the-box, session/conversation-level evaluation support is on the roadmap, and if so, any anticipated delivery date. I need this and would like to know if I should have our team solve the problem independently.

**Comment by rileymiller** [comment]
+1 this would be a fantastic addition and is really something we need.

**Comment by brioAI** [comment]
This feature is very important.
Actually, it would also be great to have it in the form a role-play:

you have one agent with one prompt collection P_agent
you role-play a fake user with a prompt collection P_user
you run a langfuse experiment in which the agent interacts with the user for K turns
and at the end you use a LLM-as-a-judge eval to evaluate the quality of your agent so that you have a score S for the the prompts P_agent
it would allow to iterate an compare two prompt collections P_agent_1 and P_agent_2 when you let your agent freely interact with a user in the wild (which is the case of MOST of companies)

**Comment by Lotte-Verheyden** [comment]
A user who monitors support chat conversations would like to track how many customers were helped by the chatbot, for which he needs evaluators that only run once the session is completed. _A session is completed when the user didn't interact for more than X minutes_.

**Comment by jannikmaierhoefer** [comment]
Hi all, we published a guide on conversational n+1 evaluation using Langfuse datasets here: https://langfuse.com/guides/cookbook/example_evaluating_multi_turn_conversations
Let me know if you have feedback!

**Comment by musaabhasan** [comment]
For multi-turn evaluation, I would model the dataset item as a conversation object rather than trying to flatten turns into a single prompt/response pair.

A compact schema could look like:

- `conversation_id`
- `turns`: ordered list of `{role, content, timestamp, tool_calls, retrieved_sources}`
- `scenario_goal`
- `success_criteria`
- `expected_facts`
- `allowed_clarifications`
- `failure_modes`
- `session_end_condition`

Then evaluators can run at two levels:

1. Turn-level: citation quality, refusal behavior, answer correctness for each assistant turn.
2. Session-level: did the whole conversation solve the task, maintain context, avoid contradiction, respect policy, and recover from ambiguity?

This matters because the sum of turn scores is not the same as conversation quality. A chatbot can answer every isolated turn reasonably while still losing the user goal, forgetting constraints, or drifting from prior commitments.

For role-play experiments, I would also store the simulated user prompt/version separately from the assistant prompt/version. Otherwise it becomes difficult to compare whether an improvement came from the agent prompt or from a less challenging synthetic user.
